### DISCUSSION TITLE
Internal PID / Artisan issue or User Error?

### DISCUSSION THREAD

**Discussion by DB-stoutheart** [artisan-roaster-scope/artisan#783] [Q&A] [answered]
So I decided to scrap the Fuji PID and give the Artisan PID a shot. The Fuji PID kept disconnecting mid-roast, modbus error.  There were other issues as well. 

So after getting the internal PID running, I've been running some tests to try and pinpoint the kp, ki, kd settings.  One issue I'm having, is when I adjust the SV, the PID doesn't adjust unless I STOP then START the PID.  Another issue is the duty will just randomly spike to 100% then pull back to 0% well before even coming close the SV. Im not sure if the PID is unstable with the current release, or if this is user error. 

Here is a look at my settings.  I figured if the SV Mode was set to Manual, and I adjust the SV, then the PID should compensate up or down, correct?

![Screen Shot 2022-01-04 at 1 13 03 PM](https://user-images.githubusercontent.com/94983037/148104620-2353246f-1095-4e32-bfe7-0d68961d20ea.png)

**Comment by MAKOMO** [comment]
Did you press the "Set" button next to the SV field to activate the new SV? Alternatively you can use the SV sliders or buttons to set the SV. If the PID is ON it should automatically adjust the duty to the new target.

**Reply by DB-stoutheart** [reply]
Thank you for getting back to me Marko.  I'll try and do a video, the behavior is very, very odd.  The PID response regardless if I adjusted the SV in the Control window or via the slider was the same. I typically would start with the SV at around 250F, I'd let the program settle around that, adjust the p-i-d settings, then adjust to 380 and it the PID would maintain the previous settings of 250.  Regardless, a video would be better.  I'll try and get to that today.

**Comment by MAKOMO** [comment, marked as answer]
Please also attach your artisan-settings.aset file here (rename to artisan-settings.txt first and then drag-and-drop in a comment field) such that I can check this using a simulator.

**Reply by DB-stoutheart** [reply]
Here you go...now for whatever reason, when I turn the burner on, after turning PID on, the duty shoots straight to 100%, as it should, then within a few seconds drops right down to minimum and doesn't even come close to SV.

I'm going to try the latest Continuous Release for the heck of it.

[01_04_2021-artisan-settings.txt](https://github.com/artisan-roaster-scope/artisan/files/7816895/01_04_2021-artisan-settings.txt)

Here is the video link - https://youtu.be/GAsSYCNQVxA

**Reply by MAKOMO** [reply]
This is a funny one. Obviously the behavior is wrong, but I cannot reproduce this on v2.4.6 with your settings using a simulated BT signal that looks like yours. I see that the Duty start to drop when the RoR gets positive, but you have d=0 thus the derivative component should not have any influence on the duty signal (I checked the code on this).

You have set "P on Input" (experimental) vs the classical "P on Error" (default), however, also this setting does only has an effect if d>0.

Could you please try the same with the Continuous Build? A number of small things did change in sampling but also in the PID algorithm that hopefully resolve this issue.

**Reply by DB-stoutheart** [reply]
So the latest continuous build seems to have the same issue.  Anything else you can recommend I try?  

I did try creating a fresh setup (factory reset) and that responded the same.

Also what is the difference between p on input vs p on error?

**Reply by MAKOMO** [reply]
The value for i should be <0 !!

"P on Input" is also called "P on Measurement" and is discussed here:

http://brettbeauregard.com/blog/2017/06/introducing-proportional-on-measurement/

I suggest to use the "P on Error" default if unclear.

**Comment by DB-stoutheart** [comment]
Hi Marko,

I have yet to try what you mentioned above, my whole family caught Covid, so I've been dealing with that.  In the meantime I've been doing a lot of research on my roaster and discovered a few things about my VFD.  The VFD allows for 0-10v and 4-20mA control. This seems like it could be very useful to me, but I'm unsure where to begin.  Would this require an additional Phidget OUT module to control via Artisan?

**Comment by DB-stoutheart** [comment]
So with the latest continuous build for Artisan, the internal PID works well.  Thank you!

**Reply by MAKOMO** [reply]
Thanks for testing and reporting!
